src/coreComponents/linearAlgebra/interfaces/trilinos/TpetraMatrix.cpp

```diff
@@ -0,0 +1,795 @@
+  GEOSX_LAI_ASSERT( P.ready() );
+  GEOSX_LAI_ASSERT_EQ( P.numLocalRows(), numLocalRows() );
+  GEOSX_LAI_ASSERT_EQ( P.numLocalRows(), numLocalCols() );
+
+  // TODO: estimate num nonzeros per row? or it does not matter since matrix is recreated anyway?
+  dst.createWithLocalSize( P.numLocalCols(), P.numLocalCols(), 1, getComm() );
+
+  Tpetra::TripleMatrixMultiply::MultiplyRAP( P.unwrapped(),
+                                             true,
+                                             *m_matrix,
+                                             false,
+                                             P.unwrapped(),
+                                             false,
+                                             dst.unwrapped() );
+  dst.m_assembled = true;
+}
+
+void TpetraMatrix::gemv( real64 const alpha,
+                         TpetraVector const & x,
+                         real64 const beta,
+                         TpetraVector & y,
+                         bool useTranspose ) const
+{
+  GEOSX_LAI_ASSERT( ready() );
+  m_matrix->apply( x.unwrapped(),
+                   y.unwrapped(),
+                   useTranspose ? Teuchos::ETransp::TRANS : Teuchos::ETransp::NO_TRANS,
+                   alpha,
+                   beta );
+}
+
+void TpetraMatrix::scale( real64 const scalingFactor )
+{
+  GEOSX_LAI_ASSERT( ready() );
+
+  if( isEqual( scalingFactor, 1.0 ) )
+  {
+    return;
+  }
+
+  open();
+  m_matrix->scale( scalingFactor );
+  close();
+}
+
+void TpetraMatrix::leftScale( TpetraVector const & vec )
+{
+  GEOSX_LAI_ASSERT( ready() );
+  m_matrix->leftScale( vec.unwrapped() );
+}
+
+void TpetraMatrix::rightScale( TpetraVector const & vec )
+{
+  GEOSX_LAI_ASSERT( ready() );
+  m_matrix->rightScale( vec.unwrapped() );
+}
+
+void TpetraMatrix::leftRightScale( TpetraVector const & vecLeft,
+                                   TpetraVector const & vecRight )
+{
+  leftScale( vecLeft );
+  rightScale( vecRight );
+}
+
+void TpetraMatrix::transpose( TpetraMatrix & dst ) const
+{
+  GEOSX_LAI_ASSERT( ready() );
+
+  using TpetraTransposer = Tpetra::RowMatrixTransposer< Tpetra_CrsMatrix::scalar_type,
+                                                        Tpetra_CrsMatrix::local_ordinal_type,
+                                                        Tpetra_CrsMatrix::global_ordinal_type >;
+
+  TpetraTransposer transposer( Teuchos::RCP< Tpetra_CrsMatrix >( m_matrix.get(), false ) );
+  Teuchos::RCP< Tpetra_CrsMatrix > trans = transposer.createTranspose();
+
+  dst.reset();
+  dst.m_matrix.reset( trans.release().get() );
+  dst.m_src_map = std::make_unique< Tpetra_Map >( *dst.m_matrix->getDomainMap() );
+  dst.m_dst_map = std::make_unique< Tpetra_Map >( *dst.m_matrix->getRangeMap() );
+  dst.m_assembled = true;
+}
+
+real64 TpetraMatrix::clearRow( globalIndex const globalRow,
+                               bool const keepDiag,
+                               real64 const diagValue )
+{
+  // TODO: Deprecate/remove this method?
```

Comment:
Is this method still used to impose BC?

Reply:
No, since BCs are now imposed on the local matrix/rhs objects prior to creation of parallel matrix/rhs.

Reply:
Ok, so it can be safely removed.